src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -284,6 +298,82 @@ private class ScopeTreeBuilder : BoundTreeWalkerWithStackGuardWithoutRecursionOn
                 private readonly HashSet<MethodSymbol> _methodsConvertedToDelegates;
                 private readonly DiagnosticBag _diagnostics;
 
+                private static readonly ObjectPool<List<Scope>> s_scopeListObjectPool = new ObjectPool<List<Scope>>(() => new List<Scope>());
+
+                /// <summary>
+                /// For every label visited so far, this dictionary maps to a list of all scopes either visited so far, or currently being visited,
+                /// that are both after the label, and are on the same level of the scope tree as the label.
+                /// </summary>
+                private readonly PooledDictionary<LabelSymbol, List<Scope>> _scopesAfterLabel = PooledDictionary<LabelSymbol, List<Scope>>.GetInstance();
```

Comment:
You can replace all of these `List`s with `ArrayBuilder` and then delete the object pool additions.

Reply:
Sure. I'm keeping the Stack for now - although it could be replaced with an ArrayBuilder without loss of performance, I believe using a stack provides immediate insight into to how it's used. Does that make sense?

Reply:
We use ArrayBuilder for stack in a lot of places, we should keep with existing convention.